File: songdata.py
from dataclasses import dataclass, field
from typing import List, Dict, Any

@dataclass(frozen=True)
class SongData:
    """
    Represents the static, immutable data for a single song, loaded from JSON.
    """
    song_id: str
    title: str
    difficulty: str
    group: str
    attribute: str
    notes: List[Dict[str, Any]] = field(default_factory=list)

File: simulator_rs/src/song/song_data.rs
use crate::song::note::Note;
use serde::Deserialize;

#[derive(Clone, Debug, Deserialize, PartialEq)]
pub struct SongData {
    #[serde(skip_deserializing)]
    pub song_id: String,
    pub title: String,
    pub difficulty: String,
    pub group: String,
    pub attribute: String,
    #[serde(default)]
    pub notes: Vec<Note>,
}
